substrate/client/transaction-pool/src/fork_aware_txpool/view.rs

```diff
@@ -0,0 +1,405 @@
+	/// terminated by sending a message to the `rx` channel provided within
+	/// `finish_revalidation_worker_channels`. Revalidation results are sent back over the `tx`
+	/// channels and shall be applied in maintain thread.
+	///
+	/// View revalidation currently is not throttled, and until not terminated it will revalidate
+	/// all the transactions. Note: this can be improved if CPU usage due to revalidation becomes a
+	/// problem.
+	pub(super) async fn revalidate(
+		&self,
+		finish_revalidation_worker_channels: FinishRevalidationWorkerChannels<ChainApi>,
+	) {
+		let FinishRevalidationWorkerChannels {
+			mut finish_revalidation_request_rx,
+			revalidation_result_tx,
+		} = finish_revalidation_worker_channels;
+
+		log::debug!(target:LOG_TARGET, "view::revalidate_later: at {} starting", self.at.hash);
+		let start = Instant::now();
+		let validated_pool = self.pool.validated_pool();
+		let api = validated_pool.api();
+
+		let batch: Vec<_> = validated_pool.ready().collect();
+		let batch_len = batch.len();
+
+		//todo: sort batch by revalidation timestamp | maybe not needed at all? xts will be getting
+		//out of the view...
+		//todo: revalidate future, remove if invalid [#5496]
+
+		let mut invalid_hashes = Vec::new();
+		let mut revalidated = HashMap::new();
+
+		let mut validation_results = vec![];
+		let mut batch_iter = batch.into_iter();
+		loop {
+			let mut should_break = false;
+			tokio::select! {
+				_ = finish_revalidation_request_rx.recv() => {
+					log::trace!(target: LOG_TARGET, "view::revalidate_later: finish revalidation request received at {}.", self.at.hash);
+					should_break = true;
+				}
+				_ = async {
+					if let Some(tx) = batch_iter.next() {
+						let validation_result = (api.validate_transaction(self.at.hash, tx.source, tx.data.clone()).await, tx.hash, tx);
+						validation_results.push(validation_result);
+					} else {
+						{
+							self.revalidation_worker_channels.lock().as_mut().map(|ch| ch.remove_sender());
+						}
+						should_break = true;
+					}
+				} => {}
+			}
+
+			if should_break {
+				break;
+			}
+		}
+
+		let revalidation_duration = start.elapsed();
+		self.metrics.report(|metrics| {
+			metrics.view_revalidation_duration.observe(revalidation_duration.as_secs_f64());
+		});
+		log::info!(
+			target:LOG_TARGET,
+			"view::revalidate_later: at {:?} count: {}/{} took {:?}",
+			self.at.hash,
+			validation_results.len(),
+			batch_len,
+			revalidation_duration
+		);
+		log_xt_debug!(data:tuple, target:LOG_TARGET, validation_results.iter().map(|x| (x.1, &x.0)), "[{:?}] view::revalidate_later result: {:?}");
+
+		for (validation_result, tx_hash, tx) in validation_results {
+			match validation_result {
+				Ok(Err(TransactionValidityError::Invalid(_))) => {
+					invalid_hashes.push(tx_hash);
+				},
+				Ok(Err(TransactionValidityError::Unknown(_))) => {
+					// skipping unknown, they might be pushed by valid or invalid transaction
+					// when latter resubmitted.
```

Comment:
What you mean by this?

Reply:
Revalidation flow was just reused, this comment was copied:
https://github.com/paritytech/polkadot-sdk/blame/a34cc8dff09dc3840a304befc2415343244b5fd0/substrate/client/transaction-pool/src/revalidation.rs#L110

I think that we want to keep transaction for which validity cannot be determined now. In some point in future this may change by some other transaction getting into the pool.

Reply:
OK, reworked as you suggested in: 39b2c3a